syntax = "proto3";

import "google/protobuf/empty.proto";

option go_package = "fastrg-controller/proto/fastrgnodepb";

package fastrgnodeservice;

service FastrgService {
  rpc ApplyConfig(ConfigRequest) returns(ConfigReply) {}
  rpc RemoveConfig(ConfigRequest) returns(ConfigReply) {}
  rpc SetSubscriberCount(SetSubscriberCountRequest) returns(SetSubscriberCountReply) {}
  rpc ConnectHsi(HsiRequest) returns(HsiReply) {}
  rpc DisconnectHsi(HsiRequest) returns(HsiReply) {}
  rpc DhcpServerStart(DhcpServerRequest) returns(DhcpServerReply) {}
  rpc DhcpServerStop(DhcpServerRequest) returns(DhcpServerReply) {}
  rpc GetFastrgSystemInfo(google.protobuf.Empty) returns(FastrgSystemInfo) {}
  rpc GetFastrgHsiInfo(google.protobuf.Empty) returns(FastrgHsiInfo) {}
  rpc GetFastrgDhcpInfo(google.protobuf.Empty) returns(FastrgDhcpInfo) {}
  rpc GetNodeStatus(google.protobuf.Empty) returns(NodeStatus) {}
}

message ConfigRequest {
  uint32 user_id = 1;
  uint32 vlan_id = 2;
  string pppoe_account = 3;
  string pppoe_password = 4;
  string dhcp_pool_start = 5;
  string dhcp_pool_end = 6;
  string dhcp_subnet_mask = 7;
  string dhcp_gateway = 8;
}

message ConfigReply { 
  string status = 1;
}

message SetSubscriberCountRequest {
  uint32 subscriber_count = 1;
}

message SetSubscriberCountReply { 
  string status = 1;
}

message HsiRequest {
  uint32 user_id = 1;
  bool force = 2;
}

message HsiReply { 
  string status = 1;
}

message DhcpServerRequest {
  uint32 user_id = 1;
}

message DhcpServerReply { 
  string status = 1;
}

message FastrgBaseInfo {
  string fastrg_version = 1;
  string build_date = 2;
  int64 uptime = 3; // in seconds
  string dpdk_version = 4;
  string dpdk_eal_args = 5;
  uint32 num_users = 6;
}

message NicDriverInfo {
  string driver_name = 1;
  string pci_addr = 2;
  bytes mac_addr = 3;
}

message PerUserStatistics {
  uint32 user_id = 1;
  uint64 rx_packets = 2;
  uint64 tx_packets = 3;
  uint64 rx_bytes = 4;
  uint64 tx_bytes = 5;
  uint64 dropped_packets = 6;
  uint64 dropped_bytes = 7;
}

message Statistics {
  uint64 rx_packets = 1;
  uint64 tx_packets = 2;
  uint64 rx_bytes = 3;
  uint64 tx_bytes = 4;
  uint64 rx_errors = 5;
  uint64 tx_errors = 6;
  uint64 rx_dropped = 7;
  repeated PerUserStatistics per_user_stats = 8;
}

message FastrgSystemInfo {
  FastrgBaseInfo base_info = 1;
  repeated NicDriverInfo nics = 2;
  repeated Statistics stats = 3;
}

message NodeStatus {
  string node_os_version = 1;
  int64 node_uptime = 2; // in seconds
  string node_ip_info = 3; // mgmt ip/netmask
  bool healthy = 4;
}

message HsiInfo {
  uint32 user_id = 1;
  uint32 vlan_id = 2;
  string status = 3;
  string account = 4;
  string password = 5;
  uint32 session_id = 6;
  string ip_addr = 7;
  string gateway = 8;
  repeated string dnss = 9;
  uint64 pppoes_rx_packets = 10;
  uint64 pppoes_tx_packets = 11;
  uint64 pppoes_rx_bytes = 12;
  uint64 pppoes_tx_bytes = 13;
}

message FastrgHsiInfo {
  repeated HsiInfo hsi_infos = 1;
}

message DhcpInfo {
  uint32 user_id = 1;
  string status = 2;
  string ip_range = 3;
  string subnet_mask = 4;
  string gateway = 5;
  repeated string inuse_ips = 6;
  repeated string dnss = 7;
}

message FastrgDhcpInfo {
  repeated DhcpInfo dhcp_infos = 1;
}
